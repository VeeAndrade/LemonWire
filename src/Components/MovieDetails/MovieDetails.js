import React, { Component } from 'react';
import './MovieDetails.css';

class MovieDetails extends Component {
  constructor() {
    super();
    this.state =
      {
        id: 21,
        title: "Sonic the Hedgehog",
        poster_path: "https://image.tmdb.org/t/p/original//aQvJ5WPzZgYVDrxLX4R6cLJCEaQ.jpg",
        backdrop_path: "https://image.tmdb.org/t/p/original//tCUcf3oNWMW8kwAj3WC6CvIN5ah.jpg",
        release_date: "2020-02-12",
        overview: "Based on the global blockbuster videogame franchise from Sega, Sonic the Hedgehog tells the story of the worldâ€™s speediest hedgehog as he embraces his new home on Earth. In this live-action adventure comedy, Sonic and his new best friend team up to defend the planet from the evil genius Dr. Robotnik and his plans for world domination.",
        average_rating: 7
      }
  }

  render() {
    return (
      <article className='movie-details-container'>
        <img src={this.state.backdrop_path} alt='movie backdrop' className='movie-backdrop'/>
        <div className='movie-details'>
          <button className='close-button'>X</button>
          <h1 className='movie-title'>{this.state.title}</h1>
          <p>{this.state.release_date}</p>
          <p className='movie-overview'>{this.state.overview}</p>
          <div className='rating-container'>
            <button className='rating-button'>+</button>
            <p className='movie-number'>{this.state.average_rating}</p>
            <button className='rating-button'>-</button>
          </div>
        </div>
      </article>
    )
  }
}

export default MovieDetails;
