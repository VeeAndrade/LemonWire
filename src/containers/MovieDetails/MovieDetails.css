* {
  margin: 0px;
  padding: 0px;
}

.movie-details-container {
  display: flex;
  height: 70%;
  width: 90%;
  background: rgb(255,237,48);
  background: linear-gradient(0deg, rgba(255,237,48,1) 0%, rgba(255,255,255,1) 64%);
  margin-top: 110px;
  border-radius: 10px;
}

.movie-backdrop {
  border-top-left-radius: 10px;
  border-bottom-left-radius: 10px;
  height: 100%;
  width: 60%;
  object-fit: cover;
}

.movie-details {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.close-button,
.submit-rating-button {
  margin: 40% 0% 0% 670%;
  height: 30px;
  width: 30px;
  border-radius: 15px;
  border: 1px solid #000;
  background-color: #FFF;
  font-size: 100%;
  font-weight: bolder;
}

.submit-rating-button {
  margin: 4%;
  width: 150px;
  border-radius: 25px;
  background-color: #fff590;
}

.submit-rating-button:hover,
#rating-dropbox:hover {
  cursor: pointer;
}

#rating-dropbox {
  height: 30px;
  width: 70px;
  font-size: 100%;
}

.movie-title {
  font-size: 180%;
  margin: 4% 0% 3% 0%;
}

.movie-overview {
  margin-top: 5%;
  width: 90%;
}

.close-button:hover,
.rating-button:hover {
  cursor: pointer;
}

.rating-container {
  display: flex;
  align-items: center;
  justify-content: center;
  margin-top: 5%;
  width: 50%;
}

.movie-number {
  margin: 0% 10% 0% 10%;
  font-size: 130%;
}
