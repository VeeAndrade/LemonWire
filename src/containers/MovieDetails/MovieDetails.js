import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import './MovieDetails.css';
import { connect } from 'react-redux';
import { getRatings } from '../../actions';

export class MovieDetails extends Component {
  constructor(props) {
    super(props);
    this.state = {ratingDropbox: null}
  }

  formatDate = (releaseDate) => {
     const monthNames = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];
    let date = releaseDate.split('-')
    return `${monthNames[date[1] - 1]} ${date[2]}, ${date[0]}`;
  }

  updateRatingState = num => {
    this.setState({ratingDropbox: parseInt(num)})
  }

  submitRating = (userId, movieId, rating) => {
    if (!this.state.ratingDropbox) {
      window.alert(`Please select a rating number to submit!`)
    } else {
      const ratingId = this.findMovieRatingId(movieId)
      if (!ratingId) {
        this.makeRatingRequest(userId, movieId, rating )
      } else {
        this.makeDeleteRequest(userId, ratingId, movieId, rating)
      }
    }
  }

  makeRatingRequest = (userId, movieId, rating) => {
    const options = {
      method: 'POST',
      body: JSON.stringify({
        movie_id: movieId,
        rating: rating
        }),
      headers: {'Content-Type': 'application/json'}
    }
    fetch(`https://rancid-tomatillos.herokuapp.com/api/v1/users/${userId}/ratings`, options)
      .then(res => {
      if(!res.ok) {
        throw Error('Something is not right, try again later')
      }
      return res.json()})
  }

  makeDeleteRequest = (userId, ratingId, movieId, rating) => {
    fetch(`https://rancid-tomatillos.herokuapp.com/api/v1/users/${userId}/ratings/${ratingId}`, {
    method: "DELETE",
    })
    .then(response => {
      if (!response.ok) {
        throw Error('Error Message');
      }
    })
    .then( () => this.getUserRatings(userId, movieId, rating ))
  }

  getUserRatings = (userId, movieId, rating ) => {
    fetch(`https://rancid-tomatillos.herokuapp.com/api/v1/users/${userId}/ratings`)
      .then(response => response.json())
      .then(ratings => this.props.addUserRatings(ratings.ratings))
      .then(() => this.makeRatingRequest(userId, movieId, rating ))
      .catch(error => console.log(error))
  }

  findMovieRatingId = id => {
    let ratingIdNum = this.props.ratings.filter(movie => movie.movie_id === id)
    if (ratingIdNum.length) {
      return ratingIdNum[0].id
    } else {
      return 0;
    }
  }

  render() {
    return (
      <article className='movie-details-container'>
        <img src={this.props.selectedMovie.movie.backdrop_path} alt='movie backdrop' className='movie-backdrop'/>
        <div className='movie-details'>
          <Link to={'/'}><button className='close-button'>X</button></Link>
          <h1 className='movie-title'>{this.props.selectedMovie.movie.title}</h1>
          <p>{this.formatDate(this.props.selectedMovie.movie.release_date)}</p>
          <p className='movie-overview'>{this.props.selectedMovie.movie.overview}</p>
          <div className='rating-container'>
            <select onChange={e => this.updateRatingState(e.target.value)} id='rating-dropbox'>
              <option value={null}>rate!</option>
              <option value={1}>1</option>
              <option value={2}>2</option>
              <option value={3}>3</option>
              <option value={4}>4</option>
              <option value={5}>5</option>
              <option value={6}>6</option>
              <option value={7}>7</option>
              <option value={8}>8</option>
              <option value={9}>9</option>
              <option value={10}>10</option>
            </select>
            <button className='submit-rating-button' onClick={() => this.submitRating(this.props.user.id, this.props.selectedMovie.movie.id, this.state.ratingDropbox)}>submit rating</button>
          </div>
          <p className='movie-number'>average rating: {this.props.selectedMovie.movie.average_rating}</p>
        </div>
      </article>
    )
  }
}

export const mapStateToProps = (state) => ({
  selectedMovie: state.selectedMovieReducer,
  user: state.user,
  ratings: state.ratings
})

export const mapDispatchToProps = dispatch => ({
  addUserRatings: ratings => dispatch(getRatings(ratings))
})

export default connect(mapStateToProps, mapDispatchToProps)(MovieDetails);
