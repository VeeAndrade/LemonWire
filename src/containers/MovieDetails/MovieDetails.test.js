import React from 'react';
import ReactDOM from 'react-dom';
import MovieDetails from './MovieDetails';
import { shallow } from 'enzyme';

describe('MovieDetails', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<MovieDetails />)
  })

  it ('should render a MovieDetails component', () => {
    expect(wrapper).toMatchSnapshot()
  })

})
