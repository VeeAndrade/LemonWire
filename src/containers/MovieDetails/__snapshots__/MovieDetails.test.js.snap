// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails should render a MovieDetails component 1`] = `undefined`;
